orttraining/orttraining/core/optimizer/bias_dropout_fusion.cc

```diff
@@ -0,0 +1,178 @@
+    if (next_node_itr == node.OutputNodesEnd()) {
+      continue;
+    }
+
+    const Node& next_node = (*next_node_itr);
+    if (!(graph_utils::IsSupportedOptypeVersionAndDomain(next_node, "Dropout", {12}, kOnnxDomain) ||
+          graph_utils::IsSupportedOptypeVersionAndDomain(next_node, "TrainableDropout", {9}, kOnnxDomain)) ||
+        next_node.GetExecutionProviderType() != node.GetExecutionProviderType()) {
+      continue;
+    }
+
+    if (!graph.GetNodeOutputsInGraphOutputs(node).empty()) {
+      continue;
+    }
+
+    Node& dropout_node = const_cast<Node&>(next_node);
+    nodes_to_fuse.push_back(dropout_node);
+
+    dropout_output.push_back(dropout_node.MutableOutputDefs()[0]);
+    dropout_output.push_back(dropout_node.MutableOutputDefs()[1]);
+
+    bool is_onnx_dropout = dropout_node.OpType().compare("Dropout") == 0;
+
+    // matching for residual Add node
+    bool has_residual_add = false;
+    for (auto last_node_itr = dropout_node.OutputNodesBegin(); last_node_itr != dropout_node.OutputNodesEnd(); ++last_node_itr) {
+      const Node& last_node = (*last_node_itr);
+
+      if (graph_utils::IsSupportedOptypeVersionAndDomain(last_node, "Add", {7}) &&
+          last_node.GetExecutionProviderType() == node.GetExecutionProviderType()) {
+        const TensorShapeProto* input1_shape = last_node.InputDefs()[0]->Shape();
+        const TensorShapeProto* input2_shape = last_node.InputDefs()[1]->Shape();
+
+        if (input1_shape == nullptr ||
+            input2_shape == nullptr ||
+            input1_shape->dim_size() < 1 ||
+            input2_shape->dim_size() < 1 ||
+            input1_shape->dim_size() != input2_shape->dim_size()) {
+          continue;
+        }
+
+        // Inputs of Residual Add must match in shape
+        bool match = true;
+        for (int i = 0; i < input1_shape->dim_size(); ++i) {
+          match &= ONNX_NAMESPACE::operator==(input1_shape->dim(i), input2_shape->dim(i));
+        }
+        if (!match) {
+          continue;
+        }
+
+        // dropout's output is not part of of graph output
+        if (!graph.GetNodeOutputsInGraphOutputs(dropout_node).empty()) {
+          continue;
+        }
+
+        Node& residual_add_node = const_cast<Node&>(last_node);
+        const std::string& dropout_output_name = dropout_node.OutputDefs()[0]->Name();
+        if (dropout_output_name == residual_add_node.InputDefs()[0]->Name()) {
+          dropout_input.push_back(residual_add_node.MutableInputDefs()[1]);  // residual
+        } else if (dropout_output_name == residual_add_node.InputDefs()[1]->Name()) {
+          dropout_input.push_back(residual_add_node.MutableInputDefs()[0]);  // residual
+        }
+
+        dropout_output[0] = residual_add_node.MutableOutputDefs()[0];
+
+        nodes_to_fuse.push_back(residual_add_node);
+        has_residual_add = true;
+        break;
+      }
+    }
+
+    if (!has_residual_add) {
+      NodeArg& dummy = graph.GetOrCreateNodeArg("", nullptr);
```

Comment:
is it necessary to insert a dummy NodeArg?

Reply:
thought inputs could be optional somehow

---
In reply to: [441038792](https://github.com/microsoft/onnxruntime/pull/4167#discussion_r441038792) [](ancestors = 441038792)

Reply:
yes... it's still needed... 
the input is optional but still need a place holder 

---
In reply to: [441039998](https://github.com/microsoft/onnxruntime/pull/4167#discussion_r441039998) [](ancestors = 441039998,441038792)